grammars/silver/compiler/extension/records/Records.sv.md

```diff
@@ -0,0 +1,370 @@
+this::AGDcl ::= quals::NTDeclQualifiers  'record'  name::IdUpper_t
+  modifiers::NonterminalModifiers  '{'  fields::RecordFields0  '}'  ';'
+{
+  this.unparse =
+    s"${quals.unparse}record ${name.lexeme} ${modifiers.unparse} {${fields.unparse}};";
+
+  local ntDcl::AGDcl =
+    nonterminalDcl(quals, 'nonterminal',
+      nameIdUpper(name, location=this.location),
+      botlNone(location=this.location), modifiers, ';', location=this.location);
+  local ntType::TypeExpr =
+    nominalTypeExpr(qNameTypeId(name, location=this.location),
+      location=this.location);
+
+  -- PARTIAL: head() and tail() are safe because names are nonempty.
+  local nameChars::[String] = explode("", name.lexeme);
+  local prodName::IdLower_t = terminal(IdLower_t,
+    implode("", toLower(head(nameChars)) :: tail(nameChars)),
+    name.location);
+
+  local prodLHS::ProductionLHS = productionLHS(
+    nameIdLower(terminal(IdLower_t, "this"), location=this.location), '::',
+    ntType, location=this.location);
+  local prodSig::ProductionSignature =
+    productionSignatureNoCL(prodLHS, '::=',
+      productionRHSNil(location=this.location), location=this.location);
+  local prodDcl::AGDcl =
+    productionDcl('abstract', 'production',
+      nameIdLower(prodName, location=this.location), prodSig, 
+      productionBody('{', productionStmtsNil(location=this.location), '}',
+        location=this.location),
+      location=this.location);
+
+  forwards to appendAGDcl(ntDcl,
+    appendAGDcl(prodDcl, fields.recordFieldAGDcl, location=this.location),
+    location=this.location);
+}
+```
+
+### `without declaring` Nonterminal Modifier
+
+Note that to support multiple records with the same field names, we don't always declare a new annotation for the field.
+The `without declaring` nonterminal modifier requests that an annotation not be declared.
+It creates an entry in the `withoutDeclaring` synthesized attribute for the annotation.
+
+```silver
+synthesized attribute withoutDeclaring::[String] occurs on NonterminalModifier,
+  NonterminalModifiers, NonterminalModifierList;
+
+terminal Without_kwd   'without'   lexer classes {KEYWORD};
+terminal Declaring_kwd 'declaring' lexer classes {KEYWORD};
+
+concrete production withoutDeclaringOne
+this::NonterminalModifier ::= 'without'  'declaring'  fieldName::IdLower_t
+{
+  propagate nonterminalModifiers;
+
+  this.errors := [];
+  this.unparse = s"without declaring ${fieldName.lexeme}";
+
+  this.withoutDeclaring = [fieldName.lexeme];
+}
+
+concrete production withoutDeclaringMany
+this::NonterminalModifier ::= 'without'  'declaring'  '{' 
+  fieldNames::FieldNames0  '}'
+{
+  propagate nonterminalModifiers;
+
+  this.errors := [];
+  this.unparse = s"without declaring {${fieldNames.unparse}}";
+
+  this.withoutDeclaring = fieldNames.names;
+}
+
+aspect default production
+this::NonterminalModifier ::=
+{ this.withoutDeclaring = []; }
+```
+
+We need the propagates for `nonterminalModifiers` because it's interfering...
```

Comment:
How is this interfering?  `NonterminalModifier` is a closed nonterminal.

Reply:
Shouldn't `silver:compiler:definition:concrete_syntax` define a default equation for it then? Or is it being a monoid attribute supposed to provide one? It doesn't look like it does; removing the propagates here and running the test suite does result in a crash from a missing equation.

I suppose I should've just added a default equation rather than a workaround here...

Reply:
`silver:compiler:definition:concrete_syntax` is an optional export of `silver:compiler:definition:core`, so `noterminalModifiers` is effectively a part of the "host language" defining the `NonterminalModifier` nonterminal - thus it doesn't need a default equation, and must be defined for all `NonterminalModifier` extension productions.

FYI if an extension defines a new attribute on a closed host nonterminal without a default equation, then the problem is categorized as a modular well-definedness issue along the same lines as an extension production lacking a forward, not an interference issue.